### ISSUE TITLE
Improve discover() signature

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#137]
Since `discover()` currently returns an `Observable`, one has to write:
```javascript
sub = wot.discover(filter).subscribe(onNext, onError, onComplete);
// ...
// sub.unsubscribe(); 
```
The algorithm does not the start actual discovery process before a subscription is made, because it can be stopped/suppressed only from a subscription.

Instead, an alternative would be to use the following conventions instead of Observables:
- return a `Promise<Cancellable>`, resolved with a cancellation object when discovery has been started, rejected if discovery cannot be started
- set  up discovery, eventual result filtering, and start the discovery process
- for each item found, call a callback (only one callback function is needed to be passed)
- when cancel() is called, stop/suppress discovery in the background. No need to return a Promise here, but it could be done for case-by-case compatibility.

The signature would look like this:
```javascript
partial interface WoT {
    Promise<Cancellable> discover(DiscoveryCallback onFound, optional ThingFilter filter);
};
callback DiscoveryCallback = void (any value);
interface Cancellable {
    void cancel();
};
```
Client code would look like this:
```javascript
wot.discover(thing => { console.log("Thing found: " + thing.name); }, 
                     thingFilter)
.then(discovery => { 
    console.log("Discovery has started."); 
    // Cancel with discovery.cancel();
}).catch( err => { console.log("Discovery error: " + error.message); };
```

**Comment by zolkis** [comment]
@mkovatsc @danielpeintner @mjkoster

**Comment by zolkis** [comment]
Example for similar solution is the [`watch()`](https://w3c.github.io/web-nfc/#dom-nfc) method in the [Web NFC API](https://w3c.github.io/web-nfc/).

**Comment by 6d77** [comment]
@zolkis  wrote:
>  (only one callback function is needed to be passed)

However this comes at the price of the client not being noticed about

* errors occuring during discovery (e.g. some network connection lost during discovery; catch clause in above example is only invoked on errors during start of discovery),
* explicit end of discovery, i.e. no further `onFound` callbacks will be called.

Both may be useful to be informed about the final result of the discovery process.

**Comment by zolkis** [comment]
> However this comes at the price of the client not being noticed about

Yes, I was aware of that, but I thought it through with the following reasoning:
- the API primarily needs to express intent to start and stop discovery and provide the discovered Things
- the client code is mainly interested in the Things that are found; so we want to keep the output of discovery simple: the stream of Things that can be discovered until discovery is stopped.
- the client code is less interested in dealing with potential intermittent connectivity problems; the API implementation/runtime/underlying platform should be able to handle that
- if there is a deal-breaker communication error, in most cases that comes early before discovery is actually started, so it is enough to reject the promise with that
- when the promise resolves, the client knows the discovery process has started
- if discovery works for a while and then stops, the client cannot really do anything about it (since it does not control underlying protocol negotiations, failovers, etc), so even if it restarts discovery itself, at most we will have a poll-loop at the client code as well, not only in the implementation. It is the implementations responsibility to try to recover from errors and to restart discovery on need, in the background. For serious connectivity issues the Thing designer could provide separate Events (if supported by the server Thing), but it should be done explicitly by allowing/describing it in the TD, rather than implicitly (by default API support).
- stopping discovery can happen for 2 reasons: 1) the client stops it via the API, 2) it ends (for instance because the Thing Directory has been fetched). In either case what happens is just that Things are not reported any more via the registered callback. Since handling of discovered Things is anyway reactive in the client code, this should not be a problem. When the client is not any more interested in new Things, just stop discovery.
- by the nature of discovery one cannot _guarantee_ that we have found _all_ Things possible, so when the discovery has ended by itself is not so important; until discovery is stopped, new Things might be reported any time, so client code needs to handle that (e.g. visually displayed lists should be dynamically updated etc).

So yes, this is just one proposal, and has some compromises, but I think actually it might be better this way. It also depends how the discussion about Observables, Events etc will go. However, we could anyway discuss whether the assumptions above are valid or not.

**Comment by zolkis** [comment]
Fixed by #174.
